misc/clock.ru

```diff
@@ -17,10 +17,23 @@ class TickTock
     # filter. Filter is going to be very implemenation specific so I suggest
     # channel/subject where channel is a subset of a subject or rather a
     # subject without wildcards. I don't think there is a need for a
-    # key/value. A simple single argument should be fine.
+    # 
+    # [Are you suggesting 'subject' as an alternative to 'pattern'? I think 'pattern' is clearer...]
+    # [...or do you mean that 'subject' forces a server to test the string and decide?]
+    # 
+    # 2) Some servers might choose to implement optional extended features.
+    # For example, some servers might support WebSocket encoding semantics
+    # (i.e., text vs. binary op-codes, data compression / encoding) or even
+    # SSE variations (i.e, Base64 encoding for binary data support).
+    # 
```

Comment:
My concern is that leaving it too open or undefined means no one can count on a common implementation. Maybe it is enough too have the extended handler without the pub/sub pattern added to that. It is an addition to the basic support of WebSockets and SSE.

If you as convinced there needs to be some kind of format indicator for subscriptions then specifying a mime type is probably a more flexible and extendable approach. I'm not convinced we need to go to that extent though. Just sending a string or bytes covers the basics. Anything above that could be left up to the the developer.

Reply:
> Maybe it is enough too have the extended handler without the pub/sub pattern added to that. It is an addition to the basic support of WebSockets and SSE.

Yes, I would have thought the same... however, I can't help but suspect that a `push` specification that doesn't address pub/sub will be far less useful.

I mean, every implementation I had that required `push` would end up using pub/sub semantics (even if it's publishing to a user's private channel). This is true for scalability concerns as well as ease of use.

> My concern is that leaving it too open or undefined means no one can count on a common implementation.

I totally agree with this concern. I think a common "threshold" should be set... However, leaving things a little open will make future updates that much easier (think WebSockets 2.0 or whatever).

> If you as convinced there needs to be some kind of format indicator for subscriptions...

I don't think this is a requirement, But it would be helpful for WebSocket support.

Specifically, [WebSocket messages come in two flavors (text / binary, op-codes 1 & 2)](https://tools.ietf.org/html/rfc6455#section-5.2), which can be set on a per-message basis.

Perhaps this specific concern could be addressed (so servers don't need to make a "best guess")... but it could easily be a server specific extension.

I think anything besides `pattern` / `channel` (and maybe `as`) should be ignored by the specification (and servers should ignore any unknown elements). This allows a common pub/sub design that will work with any server as well as allow future extensions.

Reply:
I too prefer including some level of pub/sub. Lets continue.

I guess I always looked at the op-codes as being there for the lower level implementation and not for the end user. I conflated text and binary as being the same or at least bytes being the base for strings.

Reply:
I very used to C and Ruby, so I did the same for a long time before I realized my JavaScript implementation would break on binary WebSocket messages (JavaScript will create a `blob` object instead of a String, breaking my assumptions)... :-)

Reply:
My JavaScript is not strong but can't a blob be turned into a string?

Reply:
The issue is with UTF-8 encoding and all that other jazz... [it never seems properly organized](http://monsur.hossa.in/2012/07/20/utf-8-in-javascript.html).

The WebSocket RFC makes it easier by [requiring that text frames are be UTF-8 encoded text](https://tools.ietf.org/html/rfc6455#section-5.6).

This makes things much easier, especially when parsing JSON (which has to be UTF-8 valid).

I'd much rather start as a UTF-8 string and fail than start as a blob - even if it's just for convenience.

Reply:
> I just realized we already have a way of detecting a string vs bytes. The ruby class.

This is true for a local `publish` call... but it's not really accurate for external incoming data.

When the message arrives using an external service (Redis / Pusher) or process (IPC, such as pipes or Unix Sockets), the encoding data is detached from the string and often unavailable.

It get a bit more complicated when designing micro-services in different languages and a common pub/sub backend.

>  I don’t want to allow users to send op codes directly.

I agree. I don't want users messing around with internal protocol concerns. 

On the other hand, I can't see a way around the issue unless we drop support for binary messages.

Reply:
I'm not sure why Redis, IPC, or sockets has anything to do with the Ruby API for publish or for that matter, write. Ruby has type information associated with a data element so that can be used in a Ruby API. For on_message the same could be true, 2 arguments, the subject/channel as the first arg and the data whether string to byte array as the second. I could see adding a third which is the subscription object but don't think that is necessary.

Reply:
P.S.,

This backend connectivity concern is something I've been struggling to solve without much success.

Our discussions brought more clarity, so [I rewrote my ideas about how this could work](https://github.com/boazsegev/iodine/blob/master/SPEC-Websocket-Draft.md#pubsub-handling), including an API for PubSubEngines... but I haven't managed to find a design I am truly happy with.

Reply:
I'd like to keep any dependency on an external pub/sub system out of the Rack API for WebSockets and SSE. The API may use similar concepts but at least in my mind it is only between the Ruby handler and the JavaScript running a browser.

Of course I have given thoughts as to how to integrate with something like NATS or another system. My thoughts were along the lines of providing 'glue' gem that would accept NATS messages and republish using the Rack Push API. Pick a different glue depending on the pub/sub system. The other vector I explored is offering a shortcut directly from the pub/sub system into the server which would push to the JavaScript. That would require inclusion of the pub/sub client in C though so not really for a gem.

Reply:
> I'd like to keep any dependency on an external pub/sub system out of the Rack API for WebSockets and SSE.

I agree.

The API shouldn't require dependencies... but shouldn't it make the "glue" gem seamless?

It would be nice if the "glue" gem didn't require anything more than being required (and perhaps informing it of the URL for the external service).

> My thoughts were along the lines of providing 'glue' gem that would accept NATS messages and republish using the Rack Push API.

I like the "glue" terminology, though I think of it more like velcro. The API should supply hooks and the external gem could attach itself.

But that does raise the question of how to design the hooks.

> The other vector I explored is offering a shortcut directly from the pub/sub system into the server which would push to the JavaScript. That would require inclusion of the pub/sub client in C though so not really for a gem.

I think that supplying a native solution for some external systems is a cool idea. I do it with Redis.

But I also think this shouldn't be part of the specifications. IMHO, this should be considered an integrated "glue" gem more than anything.

Reply:
If there is a common server or touch point then the glue/velcro could simply use that for hooking up. Something like Rack::Push::Server or something.

Reply:
> If there is a common server or touch point then the glue/velcro could simply use that for hooking up

I think that's what I've been trying to do with `Rack::Push.pubsub_default=(engine)`, `Rack::Push.pubsub_register(engine)` and `Rack::Push.pubsub_reset(engine)`...

...but I'm not sure my design is all that great.

The idea was for `Rack::Push.pubsub_register(engine)` to add the engine to an array / list. When `subscribe` is called with a new channel, all the engines are notified. When a channel loses all it's subscriptions, again all the engines are informed (so they remove the channel).

This allows easy access to multiple pub/sub backends.

Publishing, on the other had is performed to a single backed (so the message isn't duplicated everywhere).

This allows network traffic to be rotated among different pub/sub services.

And I think the design could be improved, it's just the best I could come up with so far.

Reply:
I think we are on different pages. See my other comments. I was looking for a WebSockets and SSE addition to the Rack spec. This expands to outside Rack and encompass not only app and browser interactions but inter-Ruby interactions and beyond. I don't want to try to do that in a web server gem and I doubt the Rack folks will want to expand the Rake spec beyond app - browser interactions. I may be wrong on the later but I think it is a reasonable assumption.

Reply:
Lets see if we can keep the two separate but leave room to have them work well together.

Reply:
>  I don't want to try to do that in a web server gem and I doubt the Rack folks will want to expand the Rake spec beyond app - browser interactions. I may be wrong on the later but I think it is a reasonable assumption.

I really don't know. I don't mind following your lead on this.

On the other hand, I think that pub/sub services shouldn't be divided up by client identity - even though connections are the main clients (rather than Ruby objects).

I feel that if we do provide pub/sub, it should be inclusive and complete. I can't help but think that developers will want to observe the server's pub/sub through the application, even if it's just for logging.

If we don't provide pub/sub, that's fine. Maybe the strong coupling between Rack::Push and pub/sub is a mistake.

I brought up the notion of pub/sub because I felt that Push without pub/sub is lacking. I can't think of a good way to leverage Push without pub/sub.

> Lets see if we can keep the two separate but leave room to have them work well together.

I'm not sure how we could do that. If you have any ideas, I'm all ears (or eyes, since I'll be reading them).

Reply:
I have an idea. Let me draw a picture tomorrow and see what you think. I believe we can keep them separate but leave an nice path for you to bring them together with a separate gem. Bed time now.